lib/autoproj/cli/ci.rb

```diff
@@ -111,36 +184,60 @@ def package_cache_path(dir, pkg, fingerprint: nil, memo: {})
                 File.join(dir, pkg.name, fingerprint)
             end
 
+            def package_cache_state(dir, pkg, memo: {})
+                fingerprint = pkg.fingerprint(memo: memo)
+                path = package_cache_path(dir, pkg, fingerprint: fingerprint, memo: memo)
+
+                {
+                    'path' => path,
+                    'cached' => File.file?(path),
+                    'metadata' => File.file?("#{path}.json"),
+                    'fingerprint' => fingerprint
+                }
+            end
+
             def pull_package_from_cache(dir, pkg, memo: {})
                 fingerprint = pkg.fingerprint(memo: memo)
                 path = package_cache_path(dir, pkg, fingerprint: fingerprint, memo: memo)
-                unless File.file?(path)
-                    return [false, fingerprint]
-                end
+                return [false, fingerprint, {}] unless File.file?(path)
+
+                metadata_path = "#{path}.json"
+                metadata =
+                    if File.file?(metadata_path)
+                        JSON.parse(File.read(metadata_path))
+                    else
+                        {}
+                    end
+
+                # Do not pull packages for which we should run tests
+                tests_enabled = pkg.test_utility.enabled?
+                tests_invoked = metadata['test'] && metadata['test']['invoked']
+                return [false, fingerprint, metadata] if tests_enabled && !tests_invoked
 
                 FileUtils.mkdir_p pkg.prefix
-                result = system("tar", "xzf", path, chdir: pkg.prefix, out: '/dev/null')
-                unless result
-                    raise "tar failed when pulling cache file for #{pkg.name}"
-                end
-                [true, pkg.fingerprint(memo: memo)]
+                result = system('tar', 'xzf', path, chdir: pkg.prefix, out: '/dev/null')
+                raise "tar failed when pulling cache file for #{pkg.name}" unless result
```

Comment:
Doesn't rubocop complain about implicit exception types?

Reply:
Doesn't seem so, and I can't find a cop that does.

Reply:
Actually, just tried to add RuntimeError explicitely, and rubocop complains by default: https://docs.rubocop.org/en/latest/cops_style/#styleredundantexception

What do you think ? Should we keep it this way, or make the exception class mandatory ?

Reply:
Personally, I prefer to have the exception class explicit but if we can't setup rubocop to require that then we should probably keep the current behavior

Reply:
We can make rubocop to ignore the explicit exception, but there's nothing to enforce its presence. I agree that I would prefer using the explicit class, so let's disable the cop and do that ?

Reply:
> (and maybe define our own cop in the future for this) 

Didn`t know that was possible. Definitely agreed. :+1: